### PR TITLE
Partial imports

### PR DESCRIPTION
**Pull request by SkySkimmer**
This adds syntax `Import M(bla, bli).` to import only some names from a module. Here names may refer to global refs or syndefs (so for instance not modules or ltac definitions).

There is also `Import M(bla(..)).` to import inductive type `bla` and its constructors and eliminators (and any other inductive in the block if it's a mutual, not sure about this decision though, please comment).
Here "eliminators" means anything in the same module with the _ind _rect etc suffixes, since we don't have registered eliminators.

Overlay: https://github.com/LPCIC/coq-elpi/pull/129

TODO Require Import? -> followup PR

### CODE REVIEW

File: checker/values.ml
Comment (SkySkimmer) [inline checker/values.ml]: Why does the checker care about libobject?

Comment (SkySkimmer) [reply checker/values.ml]: cc @ppedrot

Comment (ppedrot) [reply checker/values.ml]: The checker doesn't, but it's useful for `votour` to know about the internals for better debugging.

**Comment by ejgallego** [comment]
Looks fine on a quick pass; maybe the commit headlines could be improved a bit?

**Comment by Zimmi48** [comment]
The syntax description of the `Import` and `Export` commands need to be updated but I can take care of this @SkySkimmer.

**Comment by Zimmi48** [comment]
Cf. new commit. @jfehrle: can you review it?

File: doc/sphinx/language/gallina-extensions.rst
Comment (jfehrle) [inline doc/sphinx/language/gallina-extensions.rst]:
```suggestion
   Appending a module name with a parenthesized list of names will
```

File: doc/sphinx/language/gallina-extensions.rst
Comment (jfehrle) [inline doc/sphinx/language/gallina-extensions.rst]:
```suggestion
   make only those names available with short names, not other names
```

File: doc/sphinx/language/gallina-extensions.rst
Comment (jfehrle) [inline doc/sphinx/language/gallina-extensions.rst]:
```suggestion
   defined in the module nor will it activate other features.
```

**Review by jfehrle — APPROVED** [review]
The syntax update looks OK.  A couple suggestions

File: doc/tools/docgram/fullGrammar
Comment (Zimmi48) [reply doc/tools/docgram/fullGrammar]: I did the splicing. I'll leave the rest of the comments for @SkySkimmer to handle.

Comment (jfehrle) [reply doc/tools/docgram/fullGrammar]: The splice should be done by adding the nonterminal to `common.edit_mlg` just above `(* end SPLICE *)`, then run doc_grammar.  Othewise doc_grammar will reverse the change made directly in the rst the next time it's run

Comment (Zimmi48) [reply doc/tools/docgram/fullGrammar]: @jfehrle This is what I did *of course*!

Comment (jfehrle) [reply doc/tools/docgram/fullGrammar]: That's what I expected, but when I examined the changes I didn't see the splice.  Maybe I forgot to refresh the page.  My bad.

**Comment by Zimmi48** [comment]
>  TODO Require Import?

It would be great indeed if `Require Import` / `Require Export` could behave exactly the same as `Require` + `Import` / `Require` + `Export`. Is there any reason why this isn't factorized in the parser?

**Comment by maximedenes** [comment]
It is folklore that they already don't behave the same (unfortunately).

**Comment by Zimmi48** [comment]
OK I didn't know. That's too bad.

**Comment by maximedenes** [comment]
> It would be great indeed if `Require Import` / `Require Export` could behave exactly the same as `Require` + `Import` / `Require` + `Export`. Is there any reason why this isn't factorized in the parser?

Even if it is not the case, the question of whether we want partial imports on `Require` is still open. Any opinions?

**Comment by Zimmi48** [comment]
I would prefer if it was supported (making things more consistent even if there are subtle differences), but I wouldn't block this PR for this reason.

**Comment by Zimmi48** [comment]
BTW, this:

> It is folklore that they already don't behave the same (unfortunately).

does not really answer my question:

>Is there any reason why this isn't factorized in the parser?

If the syntax are the same (despite the behavior being slightly different).

**Comment by maximedenes** [comment]
> I would prefer if it was supported (making things more consistent even if there are subtle differences), but I wouldn't block this PR for this reason.

The risk is to be a bit confusing because `Require Import Foo(mydef)` will bring all long names in `Foo` into scope. But I don't have a strong opinion.

**Comment by Zimmi48** [comment]
That's a good point. In particular, this would differ from Python's `from foo import mydef` which makes `mydef` available by its short name and nothing else. The way in Python to make the whole content of the module available by its long name, and `mydef` available by its short name would be:

```python
import foo
from foo import mydef
```

Not to say we should take inspiration from Python (the reuse of `import` for several things can also be seen as confusing).

But maybe we should not rush the decision for `Require Import`.

**Review by maximedenes — CHANGES_REQUESTED** [review]
> (and any other inductive in the block if it's a mutual, not sure about this decision though, please comment)

I think I would prefer that it imports only the blocks that are explicitly mentioned.

**Comment by SkySkimmer** [comment]
Changed:
>There is also `Import M(bla(..)).` to import inductive type `bla` and its constructors and eliminators (and any other inductive in the block if it's a mutual, not sure about this decision though, please comment).

Now only imports the inductive/constructor/eliminator for the one inductive, not the whole mutual block.

**Comment by SkySkimmer** [comment]
Overlay done, should be ready

File: library/libobject.ml
Comment (maximedenes) [inline library/libobject.ml:27]: How about naming it `unfiltered_open` instead?

**Review by maximedenes — APPROVED** [review]
LGTM modulo a nit on naming. I'm afraid this PR needs a rebase, after which I can merge it.

File: library/libobject.ml
Comment (SkySkimmer) [reply library/libobject.ml:27]: Sounds a bit too much like an unconditional open, instead of opening only when unfiltered IMO

**Comment by SkySkimmer** [comment]
@Zimmi48 the rebase had conflicts in the grammar doc stuff, I think I resolved correctly but you may want to have a look

File: library/libobject.ml
Comment (maximedenes) [reply library/libobject.ml:27]: Ok, fair enough.

**Review by Zimmi48 — APPROVED** [review]

**Comment by maximedenes** [comment]
Is the Windows failure tracked somewhere? It looks unrelated.

**Comment by Zimmi48** [comment]
It's probably #11927

**Comment by SkySkimmer** [comment]
The log clearly says `Error on dynamically loaded library: .\kernel/byterun\dllbyterun_stubs.dll: The specified module could not be found`

**Comment by Zimmi48** [comment]
Such failure already happened on master: https://dev.azure.com/coq/coq/_build/results?buildId=8839&view=logs&jobId=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&j=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&t=be5e7cb1-b8ca-569b-a02f-a5442c8a3f8a
for commit 75589e466c9f0ed62dd10fff6c1f7d284c7ffa89 while for commit 13e727ae9c7400ae0f2d9c2a5057c50e1d9b5858 it was successful, so this seems to be a random failure.

**Comment by Zimmi48** [comment]
And indeed, it happened even before that (https://dev.azure.com/coq/coq/_build/results?buildId=8826&view=logs&jobId=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&j=2d2b3007-3c5c-5840-9bb0-2b1ea49925f3&t=be5e7cb1-b8ca-569b-a02f-a5442c8a3f8a) for commit 459faeddd095789affa30e7cf3b59fdaf95c1a0f.

**Comment by Zimmi48** [comment]
In any case, I don't think it is tracked anywhere. It was probably overlooked because of the recurring Cygwin failures.

**Comment by maximedenes** [comment]
@SkySkimmer Thanks for this nice piece of work!

**Comment by Zimmi48** [comment]
As noticed by @robbertkrebbers, this was merged without a changelog.

@SkySkimmer If you strip the checklist reminding of updating doc and changelog, please consider using the `needs:` label as an alternative reminder. Otherwise, we end up with such situations where every one (both author and reviewer) forgets about the changelog.

**Comment by jfehrle** [comment]
> this was merged without a changelog.

A thought: Perhaps coqbot could be trained to check that there's a changelog entry, although for some PRs, such as doc, we generally don't create changelists.  Also seems we'd need a way to tell coqbot that no changelog is needed to handle any corner cases.  Perhaps a label could be used to indicate that?

**Comment by robbertkrebbers** [comment]
> As noticed by @robbertkrebbers, this was merged without a changelog.

It was actually [transitively noticed by @Blaisorblade](https://gitlab.mpi-sws.org/iris/iris/-/merge_requests/732#note_73464), so the credits go to him :)

**Comment by Zimmi48** [comment]
> > this was merged without a changelog.
> 
> A thought: Perhaps coqbot could be trained to check that there's a changelog entry, although for some PRs, such as doc, we generally don't create changelists. Also seems we'd need a way to tell coqbot that no changelog is needed to handle any corner cases. Perhaps a label could be used to indicate that?

Initially that was one of the option considered. But there are many PRs that are not supposed to get a changelog, so this would be a hassle. It turns out that the introduction of the PR checklist was pretty successful to help people remember about this requirement (as I proved in [my thesis](https://www.irif.fr/_media/users/theo/memoirthesis.pdf), see section 3.5.3). But of course, this doesn't work when developers strip out the template as a reflex.

**Pull request merged**
